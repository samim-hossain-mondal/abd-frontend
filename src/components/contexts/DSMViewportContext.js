import {createContext} from 'react';

const DSMViewportContext = createContext(null);

export default DSMViewportContext;