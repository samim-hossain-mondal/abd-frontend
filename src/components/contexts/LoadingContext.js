import React, { useMemo, useState } from 'react';
import PropTypes from 'prop-types';

export const LoadingContext = React.createContext();

export function LoadingProvider({ children }) {
  const [loading, setLoading] = useState(true);

  const errorContextValues = useMemo(() => ({ loading, setLoading }), [loading]);

  return (
    <LoadingContext.Provider value={errorContextValues}>
      {children}
    </LoadingContext.Provider>
  );
}

LoadingProvider.propTypes = {
  children: PropTypes.node.isRequired,
};