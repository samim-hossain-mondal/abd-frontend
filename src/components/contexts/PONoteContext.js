import React, { createContext, useState, useMemo } from 'react';
import PropTypes from 'prop-types';

export const PONoteContext = createContext();

export function PONoteProvider({ children }) {
  const [poNotes, setPONotes] = useState([]);
  const refreshContextValues = useMemo(() => ({ poNotes, setPONotes }), [poNotes]);
  return (
    <PONoteContext.Provider value={refreshContextValues}>
      {children}
    </PONoteContext.Provider>
  );
}

PONoteProvider.propTypes = {
  children: PropTypes.node.isRequired,
};