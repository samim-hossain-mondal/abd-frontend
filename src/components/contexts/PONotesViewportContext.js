import { createContext } from 'react';

const PONotesViewportContext = createContext(null);

export default PONotesViewportContext;

