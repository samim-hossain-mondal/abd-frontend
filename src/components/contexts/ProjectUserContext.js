// import axios from "axios";
import React, { createContext, useMemo, useState } from "react";
import PropTypes from "prop-types";
import axios from "axios";
import makeRequest from "../utilityFunctions/makeRequest/index";
import {
  GET_ME,
  GET_PROJECTS,
  GET_PROJECT_BY_ID,
  CREATE_PROJECT,
} from "../constants/apiEndpoints";
import { DOMAIN } from "../../config";

export const ProjectUserContext = createContext();

export function ProjectUserProvider({ children }) {
  const [projectId, setProjectId] = useState();
  const [projectDetails, setProjectDetails] = useState(undefined);

  const [user, setUser] = useState({});
  const [userRole, setUserRole] = useState();

  const [projects, setProjects] = useState([]);
  const [projectsUpdated, setProjectsUpdated] = useState(false);

  const updateUserDetails = async (setError, setSuccess) => {
    try {
      const userDetailsPromises = [];
      userDetailsPromises.push(makeRequest(GET_ME));
      userDetailsPromises.push(makeRequest(GET_PROJECTS));

      const userDetails = await Promise.all(userDetailsPromises);
      setUser(userDetails[0]);
      setProjects(userDetails[1]);
      setProjectsUpdated(true);
      setSuccess("User details loaded successfully");
    } catch (err) {
      setError(err.message);
    }
  };

  const updateProjectDetails = async (projectIdParam, setError) => {
    try {
      setProjectId(projectIdParam);
      const resData = await makeRequest(GET_PROJECT_BY_ID(projectIdParam));
      const memberData = resData.projectMembers?.find(
        (member) => member.memberId === user.memberId
      );
      if (!memberData) {
        throw new Error("Failed to know the role");
      }
      setUserRole(memberData.role);
      setProjectDetails(resData);
      return resData;
    } catch (err) {
      setError(err.message);
      return false;
    }
  };

  const addNewProject = async (title, projectDesc) => {
    const body = {
      projectName: title,
      projectDescription: projectDesc,
    };
    const { result } = await makeRequest(CREATE_PROJECT, {
      data: body,
    });
    const project = {
      projectId: result.projectId,
      projectName: result.projectName,
      projectDescription: result.projectDescription,
      _count: {
        projectMembers: result.projectMembers.length, // TODO: once the backend is fixed, remove this and use the count from the response
      },
    };
    setProjects([...projects, project]);
  };

  const fetchProjectInfo = async () => {
    const { memberId } = user;
    console.log("memberId", memberId);
    if (memberId !== null && projects.length > 0) {
      const requests = projects.map((project) =>
        axios
          .get(
            `${DOMAIN}/api/management/project/${project.projectId}/member/${memberId}`
          )
          .then((response) => {
            const projectWithRole = Object.assign(project, {
              role: response.data.role,
            });
            return projectWithRole;
          })
          .catch((error) => error)
      );

      Promise.all(requests)
        .then((updatedProjects) => {
          setProjects(updatedProjects);
        })
        .catch((error) => {
          console.log(error);
        });
    } else {
      setProjects([]);
    }
  };

  const deleteProject = async (projectIdParam) => {
    axios
      .delete(`${DOMAIN}/api/management/project/${projectIdParam}`)
      .then(() => {
        const newProjectArray = projects.filter(
          (project) => project.projectId !== projectIdParam
        );
        setProjects(newProjectArray);
      })
      .catch((error) => {
        console.log(error);
      });
  };

  const projectUserContextValues = useMemo(
    () => ({
      projectId,
      setProjectId,
      user,
      projects,
      updateUserDetails,
      projectsUpdated,
      projectDetails,
      setProjectDetails,
      updateProjectDetails,
      userRole,
      addNewProject,
      fetchProjectInfo,
      deleteProject,
    }),
    [
      projectId,
      user,
      projects,
      projectDetails,
      projectsUpdated,
      userRole,
      addNewProject,
      fetchProjectInfo,
      deleteProject,
    ]
  );

  return (
    <ProjectUserContext.Provider value={projectUserContextValues}>
      {children}
    </ProjectUserContext.Provider>
  );
}

ProjectUserProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
