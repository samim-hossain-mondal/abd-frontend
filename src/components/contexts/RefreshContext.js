import React,{createContext,useState,useMemo} from 'react';
import PropTypes from 'prop-types';

export const RefreshContext = createContext();

export function RefreshContextProvider({children}) {
  const [refresh, setRefresh] = useState({
    poNotes: false,
    sentiment: false,
    request: false,
    announcement: false,
    celebration: false,
    availabilityCalendar: false
  });
  const refreshContextValues = useMemo(() => ({ refresh, setRefresh }), [refresh]);
  return (
    <RefreshContext.Provider value = {refreshContextValues}>
      {children}
    </RefreshContext.Provider>
  );
}

RefreshContextProvider.propTypes = {
  children: PropTypes.node.isRequired,
};