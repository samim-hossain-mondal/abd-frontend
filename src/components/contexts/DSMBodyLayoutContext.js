import React, { useReducer, useMemo, createContext, useContext, useEffect } from 'react';
import {
  Grid,
} from '@mui/material';
import PropTypes from 'prop-types';
import { isMember } from '../constants/users';
import { ProjectUserContext } from './ProjectUserContext';

const intitalGridHeightState = (member = false) => ({
  sentiment: {
    height: member ? '25%' : "9%",
    expanded: member
  },
  celebration: {
    height: member ? '75%' : "91%",
    expanded: true,
    fullExpanded: false
  },
  request: {
    height: '50%',
    expanded: true
  },
  announcement: {
    height: '50%',
    expanded: true
  }
})

const gridHeightReducer = (state, action) => {
  switch (action.type) {
    case 'SENTIMENT': {
      if (!state.sentiment.expanded && state.celebration.fullExpanded) return intitalGridHeightState(isMember(action.userRole));
      if (state.sentiment.expanded) {
        return {
          ...state,
          sentiment: {
            ...state.sentiment,
            height: '9%',
            expanded: false
          },
          celebration: {
            ...state.celebration,
            height: '91%'
          }
        }
      }
      return {
        ...state,
        sentiment: intitalGridHeightState(isMember(action.userRole)).sentiment,
        celebration: {
          ...state.celebration,
          height: intitalGridHeightState(isMember(action.userRole)).celebration.height,
        }
      }
    }
    case 'REQUEST': {
      if (!state.request.expanded && state.celebration.fullExpanded) return intitalGridHeightState(isMember(action.userRole))
      if (state.request.expanded) {
        if (state.announcement.expanded) {
          return {
            ...state,
            request: {
              ...state.request,
              height: '9%',
              expanded: false
            },
            announcement: {
              ...state.announcement,
              height: '91%'
            }
          }
        }
        return {
          ...state,
          celebration: {
            ...state.celebration,
            height: '91%',
            fullExpanded: true
          },
          sentiment: {
            ...state.sentiment,
            height: '9%',
            expanded: false
          },
          request: {
            ...state.request,
            height: '9%',
            expanded: false
          },
          announcement: {
            ...state.announcement,
            height: '9%'
          },
        }
      }
      return {
        ...state,
        celebration: {
          ...state.celebration,
          fullExpanded: false
        },
        request: intitalGridHeightState(isMember(action.userRole)).request,
        announcement: intitalGridHeightState(isMember(action.userRole)).announcement
      }
    }
    case 'ANNOUNCEMENT': {
      if (!state.announcement.expanded && state.celebration.fullExpanded) return intitalGridHeightState(isMember(action.userRole))
      if (state.announcement.expanded) {
        if (state.request.expanded) {
          return {
            ...state,
            request: {
              ...state.request,
              height: '91%',
            },
            announcement: {
              ...state.announcement,
              height: '9%',
              expanded: false
            }
          }
        }
        return {
          ...state,
          celebration: {
            ...state.celebration,
            height: '91%',
            fullExpanded: true
          },
          sentiment: {
            ...state.sentiment,
            height: '9%',
            expanded: false
          },
          request: {
            ...state.request,
            height: '9%'
          },
          announcement: {
            ...state.announcement,
            height: '9%',
            expanded: false
          }
        }
      }
      return {
        ...state,
        celebration: {
          ...state.celebration,
          fullExpanded: false
        },
        request: intitalGridHeightState(isMember(action.userRole)).request,
        announcement: intitalGridHeightState(isMember(action.userRole)).announcement
      }
    }
    case 'CELEBRATION': {
      if (state.celebration.fullExpanded) {
        return intitalGridHeightState(isMember(action.userRole));
      }
      return {
        ...state,
        celebration: {
          ...state.celebration,
          height: '91%',
          fullExpanded: true
        },
        sentiment: {
          ...state.sentiment,
          height: '9%',
          expanded: false
        },
        request: {
          ...state.request,
          height: '9%',
          expanded: false
        },
        announcement: {
          ...state.announcement,
          height: '9%',
          expanded: false
        }
      }
    }
    default: return intitalGridHeightState(isMember(action.userRole));
  }
}

export const DSMBodyLayoutContext = createContext({ gridHeightState: {}, dispatchGridHeight: () => { } });

export function DSMBodyLayoutProvider({ children }) {

  const { userRole } = useContext(ProjectUserContext)
  const [gridHeightState, dispatchGridHeight] = useReducer(gridHeightReducer, intitalGridHeightState(isMember(userRole)))
  const dsmBodyLayoutContextValue = useMemo(() => ({ gridHeightState, dispatchGridHeight }), [gridHeightState])

  useEffect(() => {
    dispatchGridHeight({ userRole })
  }, [userRole])

  return (
    <DSMBodyLayoutContext.Provider value={dsmBodyLayoutContextValue}>
      <Grid sx={{ backgroundColor: '#e6eef2', height: '100%', paddingBottom: '2%' }}>
        {children}
      </Grid>
    </DSMBodyLayoutContext.Provider>
  );
}

DSMBodyLayoutProvider.propTypes = {
  children: PropTypes.node.isRequired,
};