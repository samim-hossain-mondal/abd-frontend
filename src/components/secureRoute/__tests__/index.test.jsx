import React from 'react';
import {render} from '@testing-library/react';
import { useParams, useNavigate } from 'react-router-dom';
import { useOktaAuth } from '@okta/okta-react';
import { ProjectUserProvider } from '../../contexts/ProjectUserContext';
import { ErrorProvider } from '../../contexts/ErrorContext';
import SecureRoute from '..';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(),
  useNavigate: jest.fn(),
}));

jest.mock('@okta/okta-react', () => ({
  useOktaAuth: jest.fn(),
}));

describe('SecureRoute', () => {
  it('should render', () => {
    useParams.mockReturnValue({projectId: '1'});
    useNavigate.mockReturnValue(jest.fn());
    useOktaAuth.mockReturnValue({
      authState: {
        isAuthenticated: true,
      }
    });
    
    const {container} = render(
      <ProjectUserProvider>
        <ErrorProvider>
          <SecureRoute>
            <div>Test</div>
          </SecureRoute>
        </ErrorProvider>
      </ProjectUserProvider>
    )
    expect(container).toMatchSnapshot();
  });
});