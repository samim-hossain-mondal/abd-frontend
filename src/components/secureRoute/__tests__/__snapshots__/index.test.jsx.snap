// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureRoute should render 1`] = `
<div>
  loading ...
  <div
    class="MuiBox-root css-0"
  />
</div>
`;
